File: Front/ConBridge/ConsoleTransceiver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Diagnostics;


namespace AlbaLilium.Nliois.ConBridge;


/// <summary>
/// (parent process transceiver)
/// Uses stdin and stdout of a child process for
/// inter-process communication in a standardized manner.
/// 
/// Be cautious of this for serious projects:
/// - There is a parent/child relation ship.
/// - It uses stdin, stdout.
/// Consider other approaches.
/// </summary>
public class ConsoleTransceiver {

	readonly StreamReader stdout;
	readonly StreamWriter stdin;

	/// <inheritdoc cref="ConsoleTransceiver"/>
	/// <remarks>
	/// Only takes a reference to the streams.
	/// Does not take ownership/disposes of them.
	/// </remarks>
	public ConsoleTransceiver(StreamReader stdout, StreamWriter stdin) {
		this.stdin = stdin;
		this.stdout = stdout;
	}
	
	/// <summary>
	/// Sends a raw string and returns a raw response string.
	/// Raw query may not contain new lines.
	/// </summary>
	private string RequestRaw(string query) {

		if (query.Contains('\n') || query.Contains('\r')) {
			throw new ArgumentException("Raw query may not contain new lines", nameof(query));
		}

		// Write query
		stdin.WriteLine(query);

		// Read response
		string response = stdout.ReadLine() ?? "";
		
		return response;
	}

}

File: Back/ConBridge/ConsoleTransceiver.py


class ConsoleTransceiver:
	"""
	(child process transceiver)
	Takes over stdin and stdout to reply to requests in a standardized manner.
	A way to achieve inter-process communication, with this being in a child process.
	This method is very hacky. Be cautious for serious projects and consider other approaches.
	"""

	def __init__(self):
		pass

	def take_over_until_exit(self):
		"""
		Takes over stdin and stdout until an exit command is received.
		Does not block the streams but assumes exclusive rights. 
		"""
		while True:
			received = input()
			if received == "exit":
				#TODO: Put command into some sort of enum
				break
			else:
				self._handle_received(received)

	def _handle_received(self, received: str):
		#TODO
		pass
	
	def _transmit_raw(self, response: str):
		"""Transmits a raw string. May not contain new lines."""
		if "\n" in response or "\r" in response:
			raise ValueError("Raw response string may not contain new lines")
		print(response, flush=True)
